#ifndef GENOTYPE_INVS_H
#define GENOTYPE_INVS_H

#include "utils.h"
#include "stat_tests.h"

#include "genotype.h"

void genotype_small_inv(inversion_t* inv, open_samFile_t* bam_file, IntervalTree<ext_read_t*>& candidate_reads_for_extension_itree, 
                std::unordered_map<std::string, std::pair<std::string, int> >& mateseqs_w_mapq_chr, char* contig_seq, hts_pos_t contig_len,
                stats_t& stats, config_t& config, StripedSmithWaterman::Aligner& aligner) {
    hts_pos_t inv_start = inv->start, inv_end = inv->end;

    hts_pos_t extend = stats.read_len + 20;

    // build alt allele
    /*
     * POS in VCF is the base BEFORE the inversion
     * END seems to be the base BEFORE the reference resumes - i.e., for a "clean" inversion (no deletion),POS == END, otherwise the last base deleted
     * As usual, in order to make intervals [ ), we increase the coordinates by 1
     */
    inv_start++; inv_end++;
\
    hts_pos_t alt_start = std::max(hts_pos_t(0), inv_start-extend);
    hts_pos_t alt_end = std::min(inv_end+extend, contig_len);
    int alt_lf_len = inv_start-alt_start, alt_rf_len = alt_end-inv_end;
    char* inv_seq;
    if (inv->ins_seq.empty()) {
        inv_seq = new char[inv->inv_end-inv->inv_start+1];
        strncpy(inv_seq, contig_seq+inv->inv_start, inv->inv_end-inv->inv_start);
        inv_seq[inv->inv_end-inv->inv_start] = 0;
        rc(inv_seq);
    } else {
        inv_seq = strdup(inv->ins_seq.c_str());
    }
    int inv_len = strlen(inv_seq);
    int alt_len = alt_lf_len + inv_len + alt_rf_len;
    char* alt_seq = new char[alt_len+1];
    strncpy(alt_seq, contig_seq+alt_start, alt_lf_len);
    strncpy(alt_seq+alt_lf_len, inv_seq, inv_len);
    strncpy(alt_seq+alt_lf_len+inv_len, contig_seq+inv->end, alt_rf_len);
    alt_seq[alt_len] = 0;

    hts_pos_t ref_start = std::max(hts_pos_t(0), inv_start-extend);
    hts_pos_t ref_end = std::min(inv_end+extend, contig_len);
    int ref_len = ref_end - ref_start;
    char* ref_seq = new char[ref_len+1];
    strncpy(ref_seq, contig_seq+ref_start, ref_len);
    ref_seq[ref_len] = 0;

    std::stringstream region;
    region << inv->chr << ":" << ref_start << "-" << ref_end;
    char* regions[1];
    regions[0] = strdup(region.str().c_str());

    hts_itr_t* iter = sam_itr_regarray(bam_file->idx, bam_file->header, regions, 1);

    bam1_t* read = bam_init1();

    int same = 0;
    std::vector<std::shared_ptr<bam1_t>> alt_better_seqs, ref_better_seqs;
    std::vector<bool> alt_better_seqs_isrc, ref_better_seqs_isrc;

    StripedSmithWaterman::Filter filter;
    StripedSmithWaterman::Alignment alt_aln, ref_aln;
    while (sam_itr_next(bam_file->file, iter, read) >= 0) {
        if (is_unmapped(read) || !is_primary(read)) continue;
        if (get_unclipped_end(read) < inv_start || inv_end < get_unclipped_start(read)) continue;
        if (inv_start < get_unclipped_start(read) && get_unclipped_end(read) < inv_end) continue;

        std::string seq = get_sequence(read);
        bool is_rc = false;
        if (bam_is_rev(read) && bam_is_mrev(read) && inv->end+stats.read_len/2 <= get_mate_endpos(read) || 
            !bam_is_rev(read) && !bam_is_mrev(read) && read->core.mpos <= inv->start-stats.read_len/2) {
            rc(seq);
            is_rc = true;
        }

        // align to ALT
        aligner.Align(seq.c_str(), alt_seq, alt_len, filter, &alt_aln, 0);

        // // align to REF
        aligner.Align(seq.c_str(), ref_seq, ref_len, filter, &ref_aln, 0);

        if (alt_aln.sw_score > ref_aln.sw_score) {
            alt_better_seqs.push_back(std::shared_ptr<bam1_t>(bam_dup1(read), bam_destroy1));
            alt_better_seqs_isrc.push_back(is_rc);
        } else if (alt_aln.sw_score < ref_aln.sw_score) {
            ref_better_seqs.push_back(std::shared_ptr<bam1_t>(bam_dup1(read), bam_destroy1));
            ref_better_seqs_isrc.push_back(is_rc);
        } else {
            same++;
        }

        if (alt_better_seqs.size() + ref_better_seqs.size() + same > 4 * stats.get_max_depth(inv->chr)) {
            alt_better_seqs.clear();
            ref_better_seqs.clear();
            same = 0;
            inv->sample_info.too_deep = true;
            break;
        }
    }

    std::string alt_consensus_seq, ref_consensus_seq;
    double alt_avg_score, ref_avg_score;
    double alt_stddev_score, ref_stddev_score;
    auto alt_better_reads_consistent = gen_consensus_and_find_consistent_seqs_subset(alt_seq, alt_better_seqs, alt_better_seqs_isrc, alt_consensus_seq, alt_avg_score, alt_stddev_score);
    auto ref_better_reads_consistent = gen_consensus_and_find_consistent_seqs_subset(ref_seq, ref_better_seqs, ref_better_seqs_isrc, ref_consensus_seq, ref_avg_score, ref_stddev_score);

    if (alt_consensus_seq.length() >= 2*config.min_clip_len) {
        // all we care about is the consensus sequence
        std::shared_ptr<consensus_t> alt_consensus = std::make_shared<consensus_t>(false, 0, 0, 0, alt_consensus_seq, 0, 0, 0, 0, 0, 0);
        extend_consensus_to_left(alt_consensus, candidate_reads_for_extension_itree, inv->start-stats.max_is, inv->start, contig_len, config.high_confidence_mapq, stats, mateseqs_w_mapq_chr); 
        extend_consensus_to_right(alt_consensus, candidate_reads_for_extension_itree, inv->end, inv->end+stats.max_is, contig_len, config.high_confidence_mapq, stats, mateseqs_w_mapq_chr);
        inv->sample_info.alt_lext_reads = alt_consensus->left_ext_reads;
        inv->sample_info.alt_rext_reads = alt_consensus->right_ext_reads;
        inv->sample_info.hq_alt_lext_reads = alt_consensus->hq_left_ext_reads;
        inv->sample_info.hq_alt_rext_reads = alt_consensus->hq_right_ext_reads;
        alt_consensus_seq = alt_consensus->sequence;

        delete[] alt_seq;
        alt_start = std::max(hts_pos_t(0), inv->start-stats.max_is);
        alt_lf_len = inv->start-alt_start;
        alt_end = std::min(inv->end+stats.max_is, contig_len);
        alt_rf_len = alt_end-inv->end;
        alt_len = alt_lf_len + inv_len + alt_rf_len;
        alt_seq = new char[alt_len+1];
        strncpy(alt_seq, contig_seq+alt_start, alt_lf_len);
        strncpy(alt_seq+alt_lf_len, inv_seq, inv_len);
        strncpy(alt_seq+alt_lf_len+inv_len, contig_seq+inv->end, alt_rf_len);
        alt_seq[alt_len] = 0;

        // align to ref+SV
        aligner.Align(alt_consensus_seq.c_str(), alt_seq, alt_len, filter, &alt_aln, 0);

        delete[] ref_seq;
        ref_start = std::max(hts_pos_t(0), inv->start-stats.max_is);
        ref_end = std::min(inv->end+stats.max_is, contig_len);
        ref_len = ref_end-ref_start;
        ref_seq = new char[ref_len+1];
        strncpy(ref_seq, contig_seq+ref_start, ref_len);
        ref_seq[ref_len] = 0;

        // align to ref
        aligner.Align(alt_consensus_seq.c_str(), ref_seq, ref_len, filter, &ref_aln, 0);

        inv->sample_info.ext_alt_consensus1_length = alt_consensus_seq.length();
        inv->sample_info.ext_alt_consensus1_to_ref_score = ref_aln.sw_score;
        inv->sample_info.ext_alt_consensus1_to_alt_score = alt_aln.sw_score;

        hts_pos_t alt_lf_len = inv->start-alt_start;
        int lf_aln_rlen = std::max(hts_pos_t(0), alt_lf_len - alt_aln.ref_begin);
        int rf_aln_rlen = std::max(hts_pos_t(0), alt_aln.ref_end - alt_lf_len);
        inv->left_anchor_aln->start = inv->start - lf_aln_rlen;
        inv->left_anchor_aln->end = inv->start;
        inv->left_anchor_aln->seq_len = lf_aln_rlen;
        inv->right_anchor_aln->start = inv->end;
        inv->right_anchor_aln->end = inv->end + rf_aln_rlen;
        inv->right_anchor_aln->seq_len = rf_aln_rlen;
    }

    set_bp_consensus_info(inv->sample_info.alt_bp1.reads_info, alt_better_seqs.size(), alt_better_reads_consistent, alt_avg_score, alt_stddev_score);
    set_bp_consensus_info(inv->sample_info.ref_bp1.reads_info, ref_better_seqs.size(), ref_better_reads_consistent, ref_avg_score, ref_stddev_score);

    inv->sample_info.alt_ref_equal_reads = same;

    delete[] alt_seq;
    delete[] ref_seq;
    delete[] inv_seq;
    
    free(regions[0]);
    bam_destroy1(read);
    hts_itr_destroy(iter);
}

void genotype_large_inv(inversion_t* inv, open_samFile_t* bam_file, IntervalTree<ext_read_t*>& candidate_reads_for_extension_itree, 
                std::unordered_map<std::string, std::pair<std::string, int> >& mateseqs_w_mapq_chr, char* contig_seq, hts_pos_t contig_len,
                stats_t& stats, config_t& config, StripedSmithWaterman::Aligner& aligner) {
    
    hts_pos_t sv_start = inv->start, sv_end = inv->end;

    hts_pos_t extend = stats.read_len + 20;

    // build alt allele
    /*
     * POS in VCF is the base BEFORE the inversion
     * END seems to be the base BEFORE the reference resumes - i.e., for a "clean" inversion (no deletion),POS == END, otherwise the last base deleted
     * As usual, in order to make intervals [ ), we increase the coordinates by 1
     */
    sv_start++; sv_end++;

    hts_pos_t alt_start = std::max(hts_pos_t(0), sv_start-extend);
    hts_pos_t alt_end = std::min(sv_end+extend, contig_len);
    int alt_lf_len = sv_start-alt_start, alt_rf_len = alt_end-sv_end;
    hts_pos_t inv_border_len = std::min(extend, inv->inv_end-inv->inv_start);

    char* inv_prefix = new char[inv_border_len+1];
    strncpy(inv_prefix, contig_seq+inv->inv_start, inv_border_len);
    inv_prefix[inv_border_len] = 0;

    char* inv_prefix_rc = strdup(inv_prefix);
    rc(inv_prefix_rc);

    char* inv_suffix = new char[inv_border_len+1];
    strncpy(inv_suffix, contig_seq+inv->inv_end-inv_border_len, inv_border_len);
    inv_suffix[inv_border_len] = 0;

    char* inv_suffix_rc = strdup(inv_suffix);
    rc(inv_suffix_rc);

    int alt_bp1_len = alt_lf_len + inv_border_len;
    char* alt_bp1_seq = new char[alt_bp1_len+1];
    strncpy(alt_bp1_seq, contig_seq+alt_start, alt_lf_len);
    strncpy(alt_bp1_seq+alt_lf_len, inv_suffix_rc, inv_border_len);
    alt_bp1_seq[alt_bp1_len] = 0;

    int alt_bp2_len = inv_border_len + alt_rf_len;
    char* alt_bp2_seq = new char[alt_bp2_len+1];
    strncpy(alt_bp2_seq, inv_prefix_rc, inv_border_len);
    strncpy(alt_bp2_seq+inv_border_len, contig_seq+sv_end, alt_rf_len);
    alt_bp2_seq[alt_bp2_len] = 0;

    hts_pos_t ref_bp1_start = std::max(hts_pos_t(0), sv_start-extend);
    hts_pos_t ref_bp1_end = std::min(sv_start+extend, contig_len);
    hts_pos_t ref_bp1_len = ref_bp1_end-ref_bp1_start;
    hts_pos_t ref_bp2_start = std::max(hts_pos_t(0), sv_end-extend);
    hts_pos_t ref_bp2_end = std::min(sv_end+extend, contig_len);
    hts_pos_t ref_bp2_len = ref_bp2_end-ref_bp2_start;

    char* regions[4];
    std::stringstream ss;
    ss << inv->chr << ":" << ref_bp1_start << "-" << ref_bp1_end;
    regions[0] = strdup(ss.str().c_str());
    ss.str("");
    
    ss << inv->chr << ":" << ref_bp2_start << "-" << ref_bp2_end;
    regions[1] = strdup(ss.str().c_str());
    ss.str("");

    ss << inv->chr << ":" << std::max(hts_pos_t(0), inv->inv_start-extend) << "-" << inv->inv_start+extend;
    regions[2] = strdup(ss.str().c_str());
    ss.str("");

    ss << inv->chr << ":" << std::max(hts_pos_t(0), inv->inv_end-extend) << "-" << inv->inv_end+extend;
    regions[3] = strdup(ss.str().c_str());
    ss.str("");

    hts_itr_t* iter = sam_itr_regarray(bam_file->idx, bam_file->header, regions, 4);

    bam1_t* read = bam_init1();

    int same = 0;
    std::vector<std::shared_ptr<bam1_t>> alt_bp1_better_reads, alt_bp2_better_reads, ref_bp1_better_reads, ref_bp2_better_reads;
    std::vector<bool> alt_bp1_better_reads_isrc, alt_bp2_better_reads_isrc, ref_bp1_better_reads_isrc, ref_bp2_better_reads_isrc;

    StripedSmithWaterman::Filter filter;
    StripedSmithWaterman::Alignment alt1_aln, alt2_aln, ref1_aln, ref2_aln;
    while (sam_itr_next(bam_file->file, iter, read) >= 0) {
        if (is_unmapped(read) || !is_primary(read)) continue;

        hts_pos_t read_start = get_unclipped_start(read), read_end = get_unclipped_end(read);
        if (overlap(read_start, read_end, sv_start, sv_start+1) == 0 && 
            overlap(read_start, read_end, sv_end, sv_end+1) == 0 &&
            overlap(read_start, read_end, inv->inv_start, inv->inv_start+1) == 0 &&
            overlap(read_start, read_end, inv->inv_end, inv->inv_end+1) == 0) continue;

        alt1_aln.Clear();
        alt2_aln.Clear();
        ref1_aln.Clear();
        ref2_aln.Clear();

        std::string seq = get_sequence(read);
        bool alt_is_rc = false, ref_is_rc = false;
        // if mate is outside the inversion and pointing towards it
        if (bam_is_mrev(read) && inv->end+stats.read_len/2 <= get_mate_endpos(read) ||
            !bam_is_mrev(read) && read->core.mpos <= inv->start-stats.read_len/2) {
            if (is_samestr(read)) { // and mate is in the same orientation, we need to rc
                rc(seq);
                alt_is_rc = true;
                ref_is_rc = true;
            }

            // align to ALT
            aligner.Align(seq.c_str(), alt_bp1_seq, alt_bp1_len, filter, &alt1_aln, 0);
            aligner.Align(seq.c_str(), alt_bp2_seq, alt_bp2_len, filter, &alt2_aln, 0);

            // align to REF
            aligner.Align(seq.c_str(), contig_seq+ref_bp1_start, ref_bp1_len, filter, &ref1_aln, 0);
            aligner.Align(seq.c_str(), contig_seq+ref_bp2_start, ref_bp2_len, filter, &ref2_aln, 0);
        } else if (inv->start-10 <= read->core.mpos && get_mate_endpos(read) <= inv->end+10) { // if mate is inside the inversion
            
            // if read and mate point towards each other, it means that if the inversion is true, they were RC together
            // therefore, we need to align it as it is to REF but reverse-complemented to ALT
            if (is_proper_pair(read, stats.min_is, stats.max_is)) {
                if (bam_is_mrev(read)) {
                    aligner.Align(seq.c_str(), contig_seq+ref_bp1_start, ref_bp1_len, filter, &ref1_aln, 0);
                    rc(seq);
                    aligner.Align(seq.c_str(), alt_bp2_seq, alt_bp2_len, filter, &alt2_aln, 0);
                    alt_is_rc = true;
                } else {
                    aligner.Align(seq.c_str(), contig_seq+ref_bp2_start, ref_bp2_len, filter, &ref2_aln, 0);
                    rc(seq);
                    aligner.Align(seq.c_str(), alt_bp1_seq, alt_bp1_len, filter, &alt1_aln, 0);
                    alt_is_rc = true;
                }
            } else if (is_samestr(read)) { // if both point in the same direction, the mate was RC by itself, and we can align the read as it is
                if (bam_is_rev(read)) {
                    aligner.Align(seq.c_str(), contig_seq+ref_bp2_start, ref_bp2_len, filter, &ref2_aln, 0);
                    aligner.Align(seq.c_str(), alt_bp2_seq, alt_bp2_len, filter, &alt2_aln, 0);
                } else {
                    aligner.Align(seq.c_str(), contig_seq+ref_bp1_start, ref_bp1_len, filter, &ref1_aln, 0);
                    aligner.Align(seq.c_str(), alt_bp1_seq, alt_bp1_len, filter, &alt1_aln, 0);
                }
            }
        }

        StripedSmithWaterman::Alignment& alt_aln = alt1_aln.sw_score >= alt2_aln.sw_score ? alt1_aln : alt2_aln;
        StripedSmithWaterman::Alignment& ref_aln = ref1_aln.sw_score >= ref2_aln.sw_score ? ref1_aln : ref2_aln;
        if (alt_aln.sw_score > ref_aln.sw_score) {
            if (alt1_aln.sw_score >= alt2_aln.sw_score) {
                alt_bp1_better_reads.push_back(std::shared_ptr<bam1_t>(bam_dup1(read), bam_destroy1));
                alt_bp1_better_reads_isrc.push_back(alt_is_rc);
            }
            if (alt1_aln.sw_score <= alt2_aln.sw_score) {
                alt_bp2_better_reads.push_back(std::shared_ptr<bam1_t>(bam_dup1(read), bam_destroy1));
                alt_bp2_better_reads_isrc.push_back(alt_is_rc);
            }
        } else if (alt_aln.sw_score < ref_aln.sw_score) {
            if (ref1_aln.sw_score >= ref2_aln.sw_score) {
                ref_bp1_better_reads.push_back(std::shared_ptr<bam1_t>(bam_dup1(read), bam_destroy1));
                ref_bp1_better_reads_isrc.push_back(ref_is_rc);
            } 
            if (ref1_aln.sw_score <= ref2_aln.sw_score) {
                ref_bp2_better_reads.push_back(std::shared_ptr<bam1_t>(bam_dup1(read), bam_destroy1));
                ref_bp2_better_reads_isrc.push_back(ref_is_rc);
            }
        } else {
            same++;
        }

        if (alt_bp1_better_reads.size() + alt_bp2_better_reads.size() + ref_bp1_better_reads.size() + ref_bp2_better_reads.size() + same > 4 * stats.get_max_depth(inv->chr)) {
            alt_bp1_better_reads.clear();
            alt_bp2_better_reads.clear();
            ref_bp1_better_reads.clear();
            ref_bp2_better_reads.clear();
            same = 0;
            inv->sample_info.too_deep = true;
            break;
        }
    }

    std::string alt_bp1_consensus_seq, alt_bp2_consensus_seq, ref_bp1_consensus_seq, ref_bp2_consensus_seq;
    double alt_bp1_avg_score, alt_bp2_avg_score, ref_bp1_avg_score, ref_bp2_avg_score;
    double alt_bp1_stddev_score, alt_bp2_stddev_score, ref_bp1_stddev_score, ref_bp2_stddev_score;
    auto alt_bp1_better_reads_consistent = gen_consensus_and_find_consistent_seqs_subset(alt_bp1_seq, alt_bp1_better_reads, alt_bp1_better_reads_isrc, alt_bp1_consensus_seq, alt_bp1_avg_score, alt_bp1_stddev_score);
    delete[] alt_bp1_seq;
    auto alt_bp2_better_reads_consistent = gen_consensus_and_find_consistent_seqs_subset(alt_bp2_seq, alt_bp2_better_reads, alt_bp2_better_reads_isrc, alt_bp2_consensus_seq, alt_bp2_avg_score, alt_bp2_stddev_score);
    delete[] alt_bp2_seq;

    char* ref_bp1_seq = new char[ref_bp1_len+1];
    strncpy(ref_bp1_seq, contig_seq+ref_bp1_start, ref_bp1_len);
    ref_bp1_seq[ref_bp1_len] = 0;
    auto ref_bp1_better_reads_consistent = gen_consensus_and_find_consistent_seqs_subset(ref_bp1_seq, ref_bp1_better_reads, ref_bp1_better_reads_isrc, ref_bp1_consensus_seq, ref_bp1_avg_score, ref_bp1_stddev_score);
    delete[] ref_bp1_seq;

    char* ref_bp2_seq = new char[ref_bp2_len+1];
    strncpy(ref_bp2_seq, contig_seq+ref_bp2_start, ref_bp2_len);
    ref_bp2_seq[ref_bp2_len] = 0;
    auto ref_bp2_better_reads_consistent = gen_consensus_and_find_consistent_seqs_subset(ref_bp2_seq, ref_bp2_better_reads, ref_bp2_better_reads_isrc, ref_bp2_consensus_seq, ref_bp2_avg_score, ref_bp2_stddev_score);
    delete[] ref_bp2_seq;

    if (alt_bp1_consensus_seq.length() >= 2*config.min_clip_len) {
        // all we care about is the consensus sequence
        std::shared_ptr<consensus_t> alt_bp1_consensus = std::make_shared<consensus_t>(false, 0, 0, 0, alt_bp1_consensus_seq, 0, 0, 0, 0, 0, 0);
        extend_consensus_to_left(alt_bp1_consensus, candidate_reads_for_extension_itree, inv->start-stats.max_is, inv->start, contig_len, config.high_confidence_mapq, stats, mateseqs_w_mapq_chr);
        extend_consensus_to_right(alt_bp1_consensus, candidate_reads_for_extension_itree, inv->start, inv->start+stats.max_is, contig_len, config.high_confidence_mapq, stats, mateseqs_w_mapq_chr);
        inv->sample_info.alt_lext_reads = alt_bp1_consensus->left_ext_reads;
        inv->sample_info.alt_rext_reads = alt_bp1_consensus->right_ext_reads;
        inv->sample_info.hq_alt_lext_reads = alt_bp1_consensus->hq_left_ext_reads;
        inv->sample_info.hq_alt_rext_reads = alt_bp1_consensus->hq_right_ext_reads;
        alt_bp1_consensus_seq = alt_bp1_consensus->sequence;
    }
    if (alt_bp2_consensus_seq.length() >= 2*config.min_clip_len) {
        // all we care about is the consensus sequence
        std::shared_ptr<consensus_t> alt_bp2_consensus = std::make_shared<consensus_t>(false, 0, 0, 0, alt_bp2_consensus_seq, 0, 0, 0, 0, 0, 0);
        extend_consensus_to_left(alt_bp2_consensus, candidate_reads_for_extension_itree, inv->end-stats.max_is, inv->end, contig_len, config.high_confidence_mapq, stats, mateseqs_w_mapq_chr);
        extend_consensus_to_right(alt_bp2_consensus, candidate_reads_for_extension_itree, inv->end, inv->end+stats.max_is, contig_len, config.high_confidence_mapq, stats, mateseqs_w_mapq_chr);
        inv->sample_info.alt_lext_reads += alt_bp2_consensus->left_ext_reads;
        inv->sample_info.alt_rext_reads += alt_bp2_consensus->right_ext_reads;
        inv->sample_info.hq_alt_lext_reads += alt_bp2_consensus->hq_left_ext_reads;
        inv->sample_info.hq_alt_rext_reads += alt_bp2_consensus->hq_right_ext_reads;
        alt_bp2_consensus_seq = alt_bp2_consensus->sequence;
    }
    
    inv->sample_info.ext_alt_consensus1_length = alt_bp1_consensus_seq.length();
    inv->sample_info.ext_alt_consensus2_length = alt_bp2_consensus_seq.length();

    ref1_aln.Clear();
    alt1_aln.Clear();
    if (alt_bp1_consensus_seq.length() >= 2*config.min_clip_len) {
        hts_pos_t ref_bp1_start = inv->start-alt_bp1_consensus_seq.length();
        if (ref_bp1_start < 0) ref_bp1_start = 0;
        hts_pos_t ref_bp1_end = inv->start+alt_bp1_consensus_seq.length();
        if (ref_bp1_end > contig_len) ref_bp1_end = contig_len;
        aligner.Align(alt_bp1_consensus_seq.c_str(), contig_seq+ref_bp1_start, ref_bp1_end-ref_bp1_start, filter, &ref1_aln, 0);

        hts_pos_t alt_bp1_start = std::max(hts_pos_t(0), hts_pos_t(inv->start-alt_bp1_consensus_seq.length()));
        hts_pos_t alt_bp1_end = std::min(hts_pos_t(inv->start+alt_bp1_consensus_seq.length()), contig_len);
        hts_pos_t alt_bp1_len = alt_bp1_end-alt_bp1_start;
        alt_bp1_seq = new char[alt_bp1_len+1];
        strncpy(alt_bp1_seq, contig_seq+alt_bp1_start, inv->start-alt_bp1_start);
        
        hts_pos_t ext_inv_border_len = std::min((hts_pos_t) alt_bp1_consensus_seq.length(), inv->inv_end-inv->inv_start);
        char* ext_inv_suffix_rc = new char[ext_inv_border_len+1];
        strncpy(ext_inv_suffix_rc, contig_seq+inv->inv_end-ext_inv_border_len, ext_inv_border_len);
        ext_inv_suffix_rc[ext_inv_border_len] = 0;
        rc(ext_inv_suffix_rc);
        strncpy(alt_bp1_seq+(inv->start-alt_bp1_start), ext_inv_suffix_rc, ext_inv_border_len);
        int extra_len = alt_bp1_consensus_seq.length() - ext_inv_border_len;
        if (extra_len > 0) {
            strncpy(alt_bp1_seq+(inv->start-alt_bp1_start)+ext_inv_border_len, contig_seq+inv->end, extra_len);
        }
        alt_bp1_seq[alt_bp1_len] = 0;

        aligner.Align(alt_bp1_consensus_seq.c_str(), alt_bp1_seq, alt_bp1_len, filter, &alt1_aln, 0);
        delete[] ext_inv_suffix_rc;
        delete[] alt_bp1_seq;

        hts_pos_t alt_lf_len = inv->start - alt_bp1_start;
        int lf_aln_rlen = std::max(hts_pos_t(0), alt_lf_len - alt1_aln.ref_begin);
        inv->left_anchor_aln->start = inv->start - lf_aln_rlen;
        inv->left_anchor_aln->end = inv->start;
        inv->left_anchor_aln->seq_len = lf_aln_rlen;
    }

    ref2_aln.Clear();
    alt2_aln.Clear();
    if (alt_bp2_consensus_seq.length() >= 2*config.min_clip_len) {
        hts_pos_t ref_bp2_start = inv->end-alt_bp2_consensus_seq.length();
        if (ref_bp2_start < 0) ref_bp2_start = 0;
        hts_pos_t ref_bp2_end = inv->end+alt_bp2_consensus_seq.length();
        if (ref_bp2_end > contig_len) ref_bp2_end = contig_len;
        aligner.Align(alt_bp2_consensus_seq.c_str(), contig_seq+ref_bp2_start, ref_bp2_end-ref_bp2_start, filter, &ref2_aln, 0);
        
        hts_pos_t alt_bp2_start = std::max(hts_pos_t(0), hts_pos_t(inv->end-alt_bp2_consensus_seq.length()));
        hts_pos_t alt_bp2_end = std::min(hts_pos_t(inv->end+alt_bp2_consensus_seq.length()), contig_len);
        hts_pos_t alt_bp2_len = alt_bp2_end-alt_bp2_start;
        alt_bp2_seq = new char[alt_bp2_len+1];
        
        hts_pos_t ext_inv_border_len = std::min((hts_pos_t) alt_bp2_consensus_seq.length(), inv->inv_end-inv->inv_start);
        int extra_len = alt_bp2_consensus_seq.length() - ext_inv_border_len;
        if (extra_len > 0) {
            strncpy(alt_bp2_seq, contig_seq+inv->start-extra_len, extra_len);
        } else extra_len = 0;
        char* ext_inv_prefix_rc = new char[ext_inv_border_len+1];
        strncpy(ext_inv_prefix_rc, contig_seq+inv->inv_start, ext_inv_border_len);
        ext_inv_prefix_rc[ext_inv_border_len] = 0;
        rc(ext_inv_prefix_rc);
        strncpy(alt_bp2_seq+extra_len, ext_inv_prefix_rc, ext_inv_border_len);
        strncpy(alt_bp2_seq+extra_len+ext_inv_border_len, contig_seq+inv->end, alt_bp2_end-inv->end);
        alt_bp2_seq[alt_bp2_len] = 0;

        aligner.Align(alt_bp2_consensus_seq.c_str(), alt_bp2_seq, alt_bp2_len, filter, &alt2_aln, 0);
        delete[] ext_inv_prefix_rc;
        delete[] alt_bp2_seq;

        hts_pos_t alt_rf_len = alt_bp2_end - inv->end;
        int rf_aln_rlen = std::max(hts_pos_t(0), alt_rf_len - (alt_bp2_len-alt2_aln.ref_end));
        inv->right_anchor_aln->start = inv->end;
        inv->right_anchor_aln->end = inv->end + rf_aln_rlen;
        inv->right_anchor_aln->seq_len = rf_aln_rlen;
    }

    inv->sample_info.ext_alt_consensus1_to_ref_score = ref1_aln.sw_score;
    inv->sample_info.ext_alt_consensus2_to_ref_score = ref2_aln.sw_score;
    inv->sample_info.ext_alt_consensus1_to_alt_score = alt1_aln.sw_score;
    inv->sample_info.ext_alt_consensus2_to_alt_score = alt2_aln.sw_score;

    set_bp_consensus_info(inv->sample_info.alt_bp1.reads_info, alt_bp1_better_reads.size(), alt_bp1_better_reads_consistent, alt_bp1_avg_score, alt_bp1_stddev_score);
    set_bp_consensus_info(inv->sample_info.alt_bp2.reads_info, alt_bp2_better_reads.size(), alt_bp2_better_reads_consistent, alt_bp2_avg_score, alt_bp2_stddev_score);
    set_bp_consensus_info(inv->sample_info.ref_bp1.reads_info, ref_bp1_better_reads.size(), ref_bp1_better_reads_consistent, ref_bp1_avg_score, ref_bp1_stddev_score);
    set_bp_consensus_info(inv->sample_info.ref_bp2.reads_info, ref_bp2_better_reads.size(), ref_bp2_better_reads_consistent, ref_bp2_avg_score, ref_bp2_stddev_score);

    inv->sample_info.alt_ref_equal_reads = same;

    free(regions[0]);
    free(regions[1]);

    bam_destroy1(read);
    hts_itr_destroy(iter);
}

bool is_small_inv(inversion_t* inv, stats_t& stats, config_t& config) {
    return inv->end-inv->start+inv->svlen() < stats.read_len-2*config.min_clip_len;
}

void genotype_invs(int id, std::string contig_name, char* contig_seq, int contig_len, std::vector<inversion_t*> invs,
    bcf_hdr_t* in_vcf_header, bcf_hdr_t* out_vcf_header, stats_t stats, config_t config, contig_map_t& contig_map,
    bam_pool_t* bam_pool, std::unordered_map<std::string, std::pair<std::string, int> >* mateseqs_w_mapq_chr) {

    StripedSmithWaterman::Aligner aligner(1, 4, 6, 1, false);

    int contig_id = contig_map.get_id(contig_name);
    read_mates(contig_id);

    std::vector<hts_pair_pos_t> target_ivals;
    for (inversion_t* inv : invs) {
        target_ivals.push_back({inv->start-stats.max_is, inv->start+stats.max_is});
        target_ivals.push_back({inv->end-stats.max_is, inv->end+stats.max_is});
    }
    std::vector<ext_read_t*> candidate_reads_for_extension;
    IntervalTree<ext_read_t*> candidate_reads_for_extension_itree = get_candidate_reads_for_extension_itree(contig_name, contig_len, target_ivals, bam_pool->get_bam_reader(id), candidate_reads_for_extension);

    open_samFile_t* bam_file = bam_pool->get_bam_reader(id);
    for (inversion_t* inv : invs) {
        if (is_small_inv(inv, stats, config)) {
            genotype_small_inv(inv, bam_file, candidate_reads_for_extension_itree, *mateseqs_w_mapq_chr, contig_seq, contig_len, stats, config, aligner);
        } else {
            genotype_large_inv(inv, bam_file, candidate_reads_for_extension_itree, *mateseqs_w_mapq_chr, contig_seq, contig_len, stats, config, aligner);
        }
    }

    for (ext_read_t* ext_read : candidate_reads_for_extension) delete ext_read;

    calculate_ptn_ratio(contig_name, invs, bam_file, config, stats);
    depth_filter_inv(contig_name, invs, bam_file, config, stats);

    // apply hard filters to inversions and calculate GT
    for (inversion_t* inv : invs) {
        bool fail_dp = false, fail_sr = false;
        inv->sample_info.filters.clear();
        if (inv->sample_info.left_flanking_cov > stats.get_max_depth(inv->chr) || inv->sample_info.right_flanking_cov > stats.get_max_depth(inv->chr) ||
            inv->sample_info.left_flanking_cov < stats.get_min_depth(inv->chr) || inv->sample_info.right_flanking_cov < stats.get_min_depth(inv->chr) ||
            inv->sample_info.left_anchor_cov > stats.get_max_depth(inv->chr) || inv->sample_info.right_anchor_cov > stats.get_max_depth(inv->chr)) {
            inv->sample_info.filters.push_back("ANOMALOUS_FLANKING_DEPTH");
            fail_dp = fail_sr = true;
        }
        if (inv->sample_info.alt_bp1.reads_info.consistent_reads() > stats.get_max_depth(inv->chr) 
            || inv->sample_info.alt_bp2.reads_info.consistent_reads() > stats.get_max_depth(inv->chr)) {
            inv->sample_info.filters.push_back("ANOMALOUS_SC_NUMBER");
            fail_dp = fail_sr = true;
        }

        if (inv->sample_info.alt_bp1.pairs_info.pos_max_mq < config.high_confidence_mapq || inv->sample_info.alt_bp2.pairs_info.pos_max_mq < config.high_confidence_mapq ||
            inv->sample_info.alt_bp1.pairs_info.neg_max_mq < config.high_confidence_mapq || inv->sample_info.alt_bp2.pairs_info.neg_max_mq < config.high_confidence_mapq) {
            inv->sample_info.filters.push_back("LOW_MAPQ_DISC_PAIRS");
            fail_dp = true;
        }
        if (inv->sample_info.alt_bp1.pairs_info.pairs < stats.get_min_disc_pairs_by_insertion_size(inv->svlen())/2) {
            inv->sample_info.filters.push_back("NOT_ENOUGH_DISC_PAIRS");
            fail_dp = true;
        }
        double ptn_ratio_bp1 = double(inv->sample_info.alt_bp1.pairs_info.pairs)/(inv->sample_info.alt_bp1.pairs_info.pairs+inv->sample_info.ref_bp1.pairs_info.pairs);
        double ptn_ratio_bp2 = double(inv->sample_info.alt_bp2.pairs_info.pairs)/(inv->sample_info.alt_bp2.pairs_info.pairs+inv->sample_info.ref_bp2.pairs_info.pairs);
        if (ptn_ratio_bp1 < 0.25 || ptn_ratio_bp2 < 0.25) {
            inv->sample_info.filters.push_back("LOW_PTN_RATIO");
            fail_dp = true;
        }

        if (inv->left_anchor_aln->end-inv->left_anchor_aln->start < stats.max_is/2 || inv->right_anchor_aln->end-inv->right_anchor_aln->start < stats.max_is/2) {
            inv->sample_info.filters.push_back("SHORT_ANCHOR");
            fail_sr = true;
        }
        if (inv->sample_info.ext_alt_consensus1_to_alt_score <= inv->sample_info.ext_alt_consensus1_to_ref_score || 
            !is_small_inv(inv, stats, config) && inv->sample_info.ext_alt_consensus2_to_alt_score <= inv->sample_info.ext_alt_consensus2_to_ref_score) {
            inv->sample_info.filters.push_back("LOW_ALT_CONSENSUS_SCORE");
            fail_sr = true;
        }

        if (!fail_sr || !fail_dp) {
            inv->sample_info.filters.clear();
            inv->sample_info.filters.push_back("PASS");
        }

        double pairs_ptn1 = inv->sample_info.alt_bp1.pairs_info.pairs/double(inv->sample_info.alt_bp1.pairs_info.pairs+inv->sample_info.ref_bp1.pairs_info.pairs);
        double pairs_ptn2 = inv->sample_info.alt_bp2.pairs_info.pairs/double(inv->sample_info.alt_bp2.pairs_info.pairs+inv->sample_info.ref_bp2.pairs_info.pairs);
        double pairs_ptn = fail_dp ? 0 : std::min(pairs_ptn1, pairs_ptn2);
        
        double reads_ptn1 = inv->sample_info.alt_bp1.reads_info.consistent_reads()/double(inv->sample_info.alt_bp1.reads_info.consistent_reads()+inv->sample_info.ref_bp1.reads_info.consistent_reads());
        double reads_ptn2 = reads_ptn1;
        if (inv->sample_info.alt_bp2.reads_info.computed) {
            reads_ptn2 = inv->sample_info.alt_bp2.reads_info.consistent_reads()/double(inv->sample_info.alt_bp2.reads_info.consistent_reads()+inv->sample_info.ref_bp2.reads_info.consistent_reads());
        }
        double reads_ptn = fail_sr ? 0 : std::min(reads_ptn1, reads_ptn2);

        double ptn = std::max(pairs_ptn, reads_ptn);
        inv->n_gt = 2;
        inv->sample_info.gt = (int*) malloc(2*sizeof(int));
        if (ptn >= 0.75) {
            inv->sample_info.gt[0] = bcf_gt_unphased(1);
            inv->sample_info.gt[1] = bcf_gt_unphased(1);
        } else if (ptn <= 0.25) {
            inv->sample_info.gt[0] = bcf_gt_unphased(0);
            inv->sample_info.gt[1] = bcf_gt_unphased(0);
        } else {
            inv->sample_info.gt[0] = bcf_gt_unphased(0);
            inv->sample_info.gt[1] = bcf_gt_unphased(1);
        }
    }
}

#endif // GENOTYPE_INVS_H